src/runtime_src/core/common/drv/cu_scu.c

```diff
@@ -0,0 +1,211 @@
+/* SPDX-License-Identifier: GPL-2.0 OR Apache-2.0 */
+/*
+ * Xilinx Soft CU
+ *
+ * Copyright (C) 2022 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: min.ma@xilinx.com, j.lin@xilinx.com
+ *
+ * This file is dual-licensed; you may select either the GNU General Public
+ * License version 2 or Apache License, Version 2.0.
+ */
+
+#include "xrt_cu.h"
+#include "xgq_cmd_ert.h"
+#include "zocl_drv.h"
+
+/* Control register bits and special behavior if any.
+ * Bit 0: ap_start(Read/Set). Clear by CU when ap_ready assert.
+ * Bit 1: ap_done(Read only). Clear on read.
+ * Bit 2: ap_idle(Read only).
+ * Bit 3: ap_ready(Read only). Self clear after clear ap_start.
+ * Bit 4: ap_continue(Read/Set). Self clear.
+ * Bit 5-7: Not support yet
+ */
+#define CTRL		0x0
+#define ARGS		0x4
+
+extern int kds_echo;
+
+static int scu_alloc_credit(void *core)
+{
+	struct xrt_cu_scu *scu = (struct xrt_cu_scu *)core;
+
+	return (scu->credits) ? scu->credits-- : 0;
+}
+
+static void scu_free_credit(void *core, u32 count)
+{
+	struct xrt_cu_scu *scu = (struct xrt_cu_scu *)core;
+
+	scu->credits += count;
+	if (scu->credits > scu->max_credits)
+		scu->credits = scu->max_credits;
+}
+
+static int scu_peek_credit(void *core)
+{
+	struct xrt_cu_scu *scu = (struct xrt_cu_scu *)core;
+
+	return scu->credits;
+}
+
+static void scu_xgq_start(struct xrt_cu_scu *scu, u32 *data)
+{
+	struct xgq_cmd_start_cuidx *cmd = (struct xgq_cmd_start_cuidx *)data;
+	u32 num_reg = 0;
+	u32 i = 0;
+	u32 *cu_regfile;
+
+	cu_regfile = scu->vaddr;
+	num_reg = (cmd->hdr.count - (sizeof(struct xgq_cmd_start_cuidx)
+				     - sizeof(cmd->hdr) - sizeof(cmd->data)))/sizeof(u32);
+	for (i = 0; i < num_reg; ++i) {
+		cu_regfile[i+1] = cmd->data[i];
+	}
+}
+
+static int scu_configure(void *core, u32 *data, size_t sz, int type)
```

Comment:
For other CU implementations, it could return errors like ENOSPC to let caller know this configure doesn't go through and try again.
This scu implementation doesn't have such thing so it always returns 0.